TypeScript Example (CJS)

Segmentational opened the pull request:
The following PR relates to a simple TypeScript implementation using `express@5.0`.

Additionally, example (found in `examples/typescript`) includes a `tsconfig.json` that compiles down to `ES5` JavaScript. A merge request I made previously had failed some tests relating to version limitations (node), so this should handle that case.

However, the compiled JavaScript and other generated assets are set to be tracked via VCS (`git`). It's of normal operation to usually *untrack* them via including `*.js` in the `.gitignore`. 

We can later add a `.gitignore` in the example folder itself. But I do want to see what the GitHub actions returns with. If appropriate, I'll update/discuss changes to these tests to account for the typescript example.

---

## Discussion ##

Looking at the other example folders, often they are without `README.md`s, `package.json`s, etc. etc.

While totally fine, the example this PR relates to, as well as another I have outstanding, perhaps could be exceptions. I know the package maintainers and other contributors see, almost daily, issues & questions relating to ESM, TypeScript, and even both.

Providing examples, like this one, hopefully will make everyone's life a bit easier `:D`.

Commit:
TypeScript Example (CJS)

Segmentational commented:
Nice, seems to have passed

bozzelliandrea commented on examples/typescript/index.ts:
```diff
@@ -0,0 +1,13 @@
+const app: import("express").Application = module.exports = express();
+
+app.get('/', function(request, response) {
+    console.log(request.url);
```
Hi @Segmentational, can you use the types here? like Request and Response types, also the import `import express from 'express';` instead of the require function.

Segmentational replied:
Absolutely, I can clean this up!

bozzelliandrea commented on examples/typescript/index.ts:
```diff
@@ -0,0 +1,13 @@
+const express = require("express");
+
+const app: import("express").Application = module.exports = express();
```
Why are you using this statement? you can easily set app value like this `const app = express();`

Segmentational replied:
Speaking from memory, I believe it was due to the narrowing and the [strictness](https://www.typescriptlang.org/docs/handbook/2/basic-types.html#strictness) of using the `@types/express` `npm` package. Overall, it's to my understanding there's a few issues with [`@types/expess`](https://github.com/DefinitelyTyped/DefinitelyTyped).

Again speaking from memory, I thought I had read in the documentation that `express` is 100% `http` compliant (relating to the built-in node.js library); however, there's loss of the exposed `HTTP.Server` types. Example

```ts
import express from "express";

import type HTTP from "http";
import type { Express } from "express-serve-static-core";

const Application: Express & HTTP.Server = express();
```

Does not allow compilation.

---

Regardless, I certainly can simplify

bozzelliandrea commented on examples/typescript/index.js.map:
```diff
@@ -0,0 +1 @@
+{"version":3,"file":"index.js","sourceRoot":"","sources":["index.ts"],"names":[],"mappings":";AAAA,MAAM,OAAO,GAAG,OAAO,CAAC,SAAS,CAAC,CAAC;AAEnC,MAAM,GAAG,GAAkC,MAAM,CAAC,OAAO,GAAG,OAAO,EAAE,CAAC;AAEtE,GAAG,CAAC,GAAG,CAAC,GAAG,EAAE,UAAS,OAAO,EAAE,QAAQ;IACnC,OAAO,CAAC,GAAG,CAAC,OAAO,CAAC,GAAG,CAAC,CAAC;IAEzB,QAAQ,CAAC,IAAI,CAAC,aAAa,CAAC,CAAC;AACjC,CAAC,CAAC,CAAC;AAEH,GAAG,CAAC,MAAM,CAAC,IAAI,CAAC ...
```
I believe this file should not be released, is the build result? same for the file tsconfig.build.info

Segmentational replied:
100% - I'll clean up

bozzelliandrea commented on examples/typescript/index.js:
```diff
@@ -0,0 +1,10 @@
+    console.log(request.url);
+    response.send("Hello World");
+});
+app.listen(3000);
```
Why is the index file cloned?

Commit:
Example Simplification, File Clean-Up

Commit:
Removed CommonJS TS Reference, Replaces `require` with `import`

Segmentational commented:
@bozzelliandrea, thank you for the feedback; I've since updated with your [recommended changes](https://github.com/expressjs/express/pull/4942/commits/bd67010a156cb8747f56f6954c5d1705d78bdda3), and point of replacing [`require` with `import`](https://github.com/expressjs/express/pull/4942/commits/18fa4f8d470dd47a4017533907178aa9d93bb0f4).

Sorry my response took so long, I'll try and get back sooner next time around `:D`.

P.S. I updated the `tsconfig.json` and `package.json` to align much closer to what a standard, default configuration would look like. Notice, too, the `latest` package version reference(s).

Using `latest` is to avoid (haha hopefully) any technical debt down the road with issues coming through asking to update the `typescript` example.

Segmentational committed:
Documentation - Conciseness

- Updated `README.md` to better reflect context
- Removed unnecessary instruction

Commit:
Target ES5 Version, Bad Ignore Pattern

bozzelliandrea reviewed (approved):
@Segmentational everything looks ok and clean, nice work! 😃 

Just my curiosity, what does CJS mean?

Segmentational commented:
> @Segmentational everything looks ok and clean, nice work! 😃
> 
> Just my curiosity, what does CJS mean?

@bozzelliandrea, sorry for the delayed response... I just moved halfway across the united states, so that took a little time haha.

Yeah, "`CJS`" can be seen in a few odd places in the `node` official documentation. The first example, although referenced as a file extension, can be seen here: 

- https://nodejs.org/api/modules.html#enabling

I'm unsure of the exact other places, but essentially it's a shorthand sometimes used to describe the `commonjs` javascript module type.

Segmentational commented:
Hello! 

So it's been quite some time, but I have a colleague who was looking for a simple TS example. I got curious, and remembered this PR. 

Are there any outstanding issues? It's likely the `tsconfig.json` is outdated. I can update that.

@dougwilson, are you still the maintainer here? 

One question: is this example outside the scope of the project? 

I have no issue closing this PR if there isn't interest in a typescript example.

But if there is, maybe we can reconsider a merge at some future release point.

Edit*: Last note, there's also a committed artifact, which is incorrect. the gitignore should be updated before this would get merged, but again, if there isn't any interest in typescript here -- we should just close this before any more work is done
